### ISSUE TITLE
InMemoryConnectionStore cache strategy

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#210]
I have some doubt about the strategy used by the LRU cache of `InMemoryConnectionStore`.

Currently when we try to add a new connection in the store we have this behavior : 

``` java
/**An entry can be successfully added to the cache if any of the
 * following conditions are met:
 * <ul>
 * <li>The cache's remaining capacity is greater than zero.</li>
 * <li>The cache contains at least one <em>stale</em> entry, i.e. an
 * entry that has not been accessed for at least the cache's <em>
 * expiration threshold</em> period. In such a case the least-
 * recently accessed stale entry gets evicted from the cache to make
 * place for the new entry to be added.</li>
 * </ul>
 * ...  ...
*/
```

1) This means there is some case where we will refuse to add connection because we have no expired entry. FMPV, we don't need this threshold,  I clearly prefer we always removed the last recently used connection.

2) If you really want to keep this threshold, I think we should at least reconsidered the default value. "36h of inactivity" is clearly too much. Something like 15/30 minutes seems enough.

3) Currently the default capacity size is 500 000. I made some tests using PSK : a connection take ~2,4KB in the heap. So 500 000 connection means 1,2GB of heap space. Does it not too much ? Should we change the default capacity size to something like 150 000 ? 

4) This 2 values in `InMemoryConnectionStore` should be configurable using [NetworkdConfig.MAX_ACTIVE_PEERS and NetworkdConfig.MAX_PEER_INACTIVITY_PERIOD](https://github.com/eclipse/californium/blob/master/californium-core/src/main/java/org/eclipse/californium/core/network/config/NetworkConfigDefaults.java#L56)

5) For security reason, the TLS specification advice to limit session lifetime to 24h. Does it mean we should limit the connection lifetime to 24h too ? (and make this value configurable)

**Comment by sophokles73** [comment]
I think you make some very valid points here. Thanks for bringing this up (again). Everything you say makes sense to me but I am currently lacking the experience from real use cases with this number of devices to be in a position to have an educated (other) opinion. Still, I have some comments:

> FMPV, we don't need this threshold, I clearly prefer we always removed the last recently used connection.

I think we should keep a limit in order to be able to determine, when we need to create more server instances in a clustered environment.

> "36h of inactivity" is clearly too much. Something like 15/30 minutes seems enough.

Considering a LWM2M use case where a client sleeps most of the time and only once in a while wakes up and tries to send a notification, a threshold of 15 to 30 minutes seems quite low for expiring the session. In the end we are discussing things like https://tools.ietf.org/html/draft-mavrogiannopoulos-tls-cid-00 so that a client doesn't need to re-negotiate a session everytime it wakes up, don't we?

> So 500 000 connection means 1,2GB of heap space. Does it not too much ? Should we change the default capacity size to something like 150 000 ?

I am fine with reducing it to a much lower value, e.g. 150000 as proposed by you. Again, I am lacking the experience from real world use cases to argue against it ;-)

> This 2 values in InMemoryConnectionStore should be configurable using NetworkdConfig.MAX_ACTIVE_PEERS and NetworkdConfig.MAX_PEER_INACTIVITY_PERIOD

Sounds good.

> For security reason, the TLS specification advice to limit session lifetime to 24h. Does it mean we should limit the connection lifetime to 24h too ? (and make this value configurable)

Yes, I agree. We should limit the session lifetime to a configurable value whose default should be 24h.

**Comment by sbernard31** [comment]
3) Ok  for 150 000. The idea is just to avoid out of memory with the default config.

4) Using `NetworkdConfig.MAX_ACTIVE_PEERS` and `NetworkdConfig.MAX_PEER_INACTIVITY_PERIOD` sounds finally not to be a good idea, as this is part of Californium and Scandium does not depend on Californium. I see with the last version this problem doesn't exist anymore. Those parameters are now configurable using DTLSConfig. 

2) About the threshold, you said that 15/30 minutes is low but keep in mind that the session will not be removed after 15/30 minutes. It just lost its "immunity" :).  It will be removed only if there is 150 000 more fresh connections than it. For me 36h its clearly too much. If you still think 15/30 minutes too low, which default value would you prefer ?

**Comment by sophokles73** [comment]
> Ok for 150 000. The idea is just to avoid out of memory with the default config.

Will do.

> I see with the last version this problem doesn't exist anymore. Those parameters are now configurable using DTLSConfig.

Yes, I totally forgot about that :-) So, we're good here, right? No need to change/add anything?

> About the threshold, you said that 15/30 minutes is low but keep in mind that the session will not be removed after 15/30 minutes. It just lost its "immunity" :). It will be removed only if there is 150 000 more fresh connections than it.

Yes, this is a very good point. We do not *destroy* the connection/session yet but it only becomes *eligible* for eviction. So let's start with 30 minutes and see how we fare. I will change it accordingly.

Thanks for your insights on this, @sbernard31. I really appreciate it :-)

**Comment by sophokles73** [comment]
fixed on master in e788d077f73e4794522910719981351604ac52e3
fixed on 2.0.x in 537432a37be43f4762d95edc1e93b5528939e727
